include/nbl/builtin/hlsl/bda/__ref.hlsl

```diff
@@ -0,0 +1,77 @@
+// Copyright (C) 2018-2024 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+
+#include "nbl/builtin/hlsl/cpp_compat.hlsl"
+#include "nbl/builtin/hlsl/spirv_intrinsics/core.hlsl"
+
+#ifndef _NBL_BUILTIN_HLSL_BDA_REF_INCLUDED_
+#define _NBL_BUILTIN_HLSL_BDA_REF_INCLUDED_
+
+namespace spirv
+{
+template<typename M, typename T, typename StorageClass>
+[[vk::ext_instruction(spv::OpAccessChain)]]
+vk::SpirvOpaqueType <spv::OpTypePointer,StorageClass,M> accessChain(
+    [[vk::ext_reference]] vk::SpirvOpaqueType <spv::OpTypePointer,StorageClass,T>base,
+    [[vk::ext_literal]] uint32_t index0
+);
+}
+
+namespace bda
+{
+template<typename T>
+using __spv_ptr_t = vk::SpirvOpaqueType<
+    spv::OpTypePointer,
+    vk::Literal< vk::integral_constant<uint, spv::StorageClassPhysicalStorageBuffer> >,
+    T
+>;
+
+template<typename T>
+struct __ptr;
+
+template<typename T, uint32_t alignment, bool _restrict>
+struct __base_ref
+{
+// TODO:
+// static_assert(alignment>=alignof(T));
+
+    using spv_ptr_t = uint64_t;
+    spv_ptr_t ptr;
+
+    __spv_ptr_t<T> __get_spv_ptr()
+    {
+        return nbl::hlsl::experimental::bitcast < __spv_ptr_t<T> > (ptr);
+    }
+
+    // TODO: Would like to use `spv_ptr_t` or OpAccessChain result instead of `uint64_t`
+    void __init(const spv_ptr_t _ptr)
+    {
+        ptr = _ptr;
+    }
+
+    T load()
+    {
+        return nbl::hlsl::spirv::load < T, __spv_ptr_t<T>, alignment > (__get_spv_ptr());
+    }
+
+    void store(const T val)
+    {
+        nbl::hlsl::spirv::store < T, __spv_ptr_t<T>, alignment > (__get_spv_ptr(), val);
+    }
+};
+
+template<typename T, uint32_t alignment/*=alignof(T)*/, bool _restrict = false>
```

Comment:
tile to use our `alignof` for the default

Reply:
Done

Reply:
alignof seems to be written for expressions? I used `alignment_of_v` instead

Reply:
ok